lib/logstash/plugin_mixins/validator_support/field_reference_validation_adapter.rb

```diff
@@ -0,0 +1,29 @@
+# encoding: utf-8
+
+require 'logstash/plugin_mixins/validator_support'
+
+module LogStash
+  module PluginMixins
+    module ValidatorSupport
+      field_name = /[^\[\]]+/                                     # anything but brackets
+      path_fragment = /\[#{field_name}\]/                         # bracket-wrapped field name
+      field_reference_literal = /#{path_fragment}+/               # one or more path fragments
+      embedded_field_reference = /\[#{field_reference_literal}\]/ # bracket-wrapped field reference literal
+      composite_field_reference = /#{Regexp.union(path_fragment, embedded_field_reference)}+/
+
+      # anchored pattern matching either a stand-alone field name, or a composite field reference
+      field_reference_pattern = /\A#{Regexp.union(field_name,composite_field_reference)}\z/
+
+      FieldReferenceValidationAdapter = NamedValidationAdapter.new(:field_reference) do |value|
+        break ValidationResult.failure("Expected exactly one field reference, got `#{value.inspect}`") unless value.kind_of?(Array) && value.size <= 1
+        break ValidationResult.success(nil) if value.empty? || value.first.nil?
+
+        candidate = value.first
+
+        break ValidationResult.failure("Expected a valid field reference, got `#{candidate.inspect}`") unless field_reference_pattern =~ candidate
```

Comment:
can we leverage the java FieldReference class here like we do in the core implementation instead of implementing this with regex matching?

Reply:
> can we leverage the java FieldReference class here

Not without tying to Logstash >= 6.1 where the primary entrypoint `org.logstashFieldReference#from(CharSequence)` was introduced (https://github.com/elastic/logstash/commit/95d56a025316a3ebfdc311164382f7c4a9b76c22), and even then, Logstash 6.0-6.3 had issues with `ArrayIndexOutOfBounds` exceptions that blew through error-handling on certain inputs (https://github.com/elastic/logstash/issues/9524), and didn't really perform _any_ validation. 

Since this `validator_support` mixin is meant to be home to other validators as we introduce them to core, I didn't see much benefit in blocking specific versions of Logstash to support this one specific validator.

As is, I didn't have to go out of my way to make this work in LS 5, so I included LS 5 in dependency declaration (and meant to include it in the matrix, but I missed that).

Reply:
As posted through another medium:


IMO the mixin helps us avoid backporting bug fixes to other Logstash
versions, typically current_minor-1, and allows those bug fixes in newer
minors to be used in any commercially supported version of Logstash
(currently 6.8+)
In this mindset I’d rather pin the mixin to 6.8+
I think that’s a good balance of broadness of compatibility without
requiring legwork like reimplementing FieldReference.from